File: src/main/kotlin/org/kravbank/java/api/RequirementVariantResource.java
package org.kravbank.java.api;


import org.kravbank.java.model.RequirementVariant;
import javax.transaction.Transactional;
import javax.ws.rs.*;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import java.net.URI;
import java.util.List;

@Path("/requirementvariants")
//@Produces(APPLICATION_JSON)
//@Consumes(APPLICATION_JSON)

public class RequirementVariantResource {
    @GET
    @Produces(MediaType.APPLICATION_JSON)
    public Response getAll() {
        List<RequirementVariant> requirementvariants = RequirementVariant.listAll();
        return Response.ok(requirementvariants).build();
    }

    @GET
    @Path("{id}")
    @Produces(MediaType.APPLICATION_JSON)
    public Response getOne(@PathParam("id") Long id) {
        return RequirementVariant.findByIdOptional(id)
                .map(requirement -> Response.ok(requirement).build())
                .orElse(Response.status(Response.Status.NOT_FOUND).build());
    }

    @POST
    @Transactional
    @Produces(MediaType.APPLICATION_JSON)
    @Consumes(MediaType.APPLICATION_JSON)
    public Response create(RequirementVariant requirement) {
        RequirementVariant.persist((requirement));
        if (requirement.isPersistent()) {
            return Response.created(URI.create("/requirementvariants" + requirement.id)).build();
        }
        return Response.status(Response.Status.BAD_REQUEST).build();
    }


    @GET
    @Path ("requirementvariant/{useproduct}")
    @Produces(MediaType.APPLICATION_JSON)
    public Response getByUseProduct(@PathParam("useproduct") String useproduct) {
        boolean useProductBool = useproduct.contentEquals("true");
        List <RequirementVariant> requirementvariants = RequirementVariant.list("SELECT p from RequirementVariant p WHERE p.useProduct = ?1 ORDER BY id DESC", useProductBool);
        return Response.ok(requirementvariants).build();
    }


    /*
    @GET
    @Path ("requirement/{needid}")
    @Produces(MediaType.APPLICATION_JSON)
    public Response getByBankId(@PathParam("needid") String needid) {
        return RequirementVariant.find("needId", needid)
                .singleResultOptional()
                .map(requirement -> Response.ok(requirement).build())
                .orElse(Response.status(Response.Status.NOT_FOUND).build());
    }
    */

    @DELETE
    @Path("{id}")
    @Produces(MediaType.APPLICATION_JSON)
    @Transactional
    public Response deleteById(@PathParam("id") Long id){
        boolean deleted = RequirementVariant.deleteById(id);
        if (deleted){
            return Response.noContent().build();
        }
        return Response.status(Response.Status.BAD_REQUEST).build();
    }
}

File: src/main/kotlin/org/kravbank/api/RequirementVariantResource.kt
package org.kravbank.api;

import org.kravbank.utils.form.requirementvariant.RequirementVariantForm
import org.kravbank.utils.form.requirementvariant.RequirementVariantFormUpdate
import org.kravbank.service.RequirementVariantService
import org.kravbank.utils.mapper.code.CodeUpdateMapper
import org.kravbank.utils.mapper.need.NeedMapper
import org.kravbank.utils.mapper.requirementvariant.RequirementVariantMapper
import java.net.URI
import java.util.ArrayList
import javax.transaction.Transactional;
import javax.ws.rs.*;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import javax.enterprise.context.RequestScoped

@Path("/api/v1/projects/{projectRef}/requirements/{requirementRef}/requirementvariants")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
@RequestScoped
class RequirementVariantResource(val requirementVariantService: RequirementVariantService) {
    //GET REQUIREMENT VARIANT
    @GET
    @Path("/{requirementVariantRef}")
    fun getRequirementVariant(
        @PathParam("projectRef") projectRef: String,
        @PathParam("requirementRef") requirementRef: String,
        @PathParam("requirementVariantRef") requirementVariantRef: String
    ): Response {
        val reqVariant = requirementVariantService.get(projectRef, requirementRef, requirementVariantRef)
        val reqVariantDTO = RequirementVariantMapper().fromEntity(reqVariant)
        return Response.ok(reqVariantDTO).build()
    }

    //LIST REQUIREMENT VARIANTS
    @GET
    fun listRequirementVariants(
        @PathParam("projectRef") projectRef: String,
        @PathParam("requirementRef") requirementRef: String,
    ): Response {
        val reqVariants = requirementVariantService.list(projectRef, requirementRef)
        val reqVariantsDTO = ArrayList<RequirementVariantForm>()
        for (c in reqVariants) reqVariantsDTO.add(RequirementVariantMapper().fromEntity(c))
        return Response.ok(reqVariantsDTO).build()
    }

    //CREATE REQUIREMENT VARIANT
    @Transactional
    @POST
    fun createRequirementVariant(
        @PathParam("projectRef") projectRef: String,
        @PathParam("requirementRef") requirementRef: String,
        newReqVariant: RequirementVariantForm
    ): Response {
        val reqVariant = requirementVariantService.create(projectRef, requirementRef, newReqVariant)
        //sender ny req variant ref i response header
        return Response.created(URI.create("/api/v1/projects/$projectRef/requirements/$requirementRef/requirementvariants/" + reqVariant.ref))
            .build()
    }

    //DELETE REQUIREMENT VARIANT
    @DELETE
    @Path("/{requirementVariantRef}")
    @Transactional
    fun deleteRequirementVariant(
        @PathParam("projectRef") projectRef: String,
        @PathParam("requirementRef") requirementRef: String,
        @PathParam("requirementVariantRef") requirementVariantRef: String
    ): Response {
        val reqvariant = requirementVariantService.delete(projectRef, requirementRef, requirementVariantRef)
        val reqVariantDTO = RequirementVariantMapper().fromEntity(reqvariant)
        // sender slettet req variant ref i body
        return Response.ok(reqVariantDTO.ref).build()
    }

    //UPDATE REQUIREMENT VARIANT
    @PUT
    @Path("{requirementVariantRef}")
    @Transactional
    fun updateRequirementVariant(
        @PathParam("projectRef") projectRef: String,
        @PathParam("requirementRef") requirementRef: String,
        @PathParam("requirementVariantRef") requirementVariantRef: String,
        updatedReqVariant: RequirementVariantFormUpdate
    ): Response {
        val reqVariant =
            requirementVariantService.update(projectRef, requirementRef, requirementVariantRef, updatedReqVariant)
        //sender DTO - mapper fra entity
        val reqVariantUpdateDTO = RequirementVariantMapper().fromEntity(reqVariant)
        return Response.ok(reqVariantUpdateDTO).build()
    }
}